frontend/src/main/java/com/google/tripmeout/frontend/storage/PersistentPlaceVisitStorage.java

```diff
@@ -0,0 +1,140 @@
+    Key placeKey = KeyFactory.stringToKey(placeVisit.id());
+    Transaction transaction = datastore.beginTransaction();
+    try {
+      datastore.get(placeKey);
+      transaction.rollback();
+      throw new PlaceVisitAlreadyExistsException("Place Already Exists");
+    } catch (EntityNotFoundException e) {
+    }
+    try {
+      Entity placeEntity = new Entity(placeKey);
+      placeEntity.setProperty(ID_PROPERTY_NAME, placeVisit.id());
+      placeEntity.setProperty(TRIP_ID_PROPERTY_NAME, placeVisit.tripId());
+      placeEntity.setProperty(USER_MARK_PROPERTY_NAME, placeVisit.userMark().toString());
+      placeEntity.setProperty(PLACE_NAME_PROPERTY_NAME, placeVisit.placeName());
+      placeEntity.setProperty(PLACE_API_ID_PROPERTY_NAME, placeVisit.placesApiPlaceId());
+      datastore.put(placeEntity);
+    } catch (Exception e) {
+      transaction.rollback();
+      throw e;
+    }
+    transaction.commit();
+  }
+
+  public void removePlaceVisit(String tripId, String placeVisitId)
+      throws PlaceVisitNotFoundException {
+    Key placeKey = KeyFactory.stringToKey(placeVisitId);
+    Transaction transaction = datastore.beginTransaction();
+    try {
+      datastore.get(placeKey);
+    } catch (EntityNotFoundException e) {
+      transaction.rollback();
+      throw new PlaceVisitNotFoundException("Place with id: " + placeVisitId + " doesn't exist");
+    }
+    datastore.delete(placeKey);
+    transaction.commit();
+  }
+
+  public Optional<PlaceVisitModel> getPlaceVisit(String tripId, String placeVisitId) {
+    Key placeKey = KeyFactory.stringToKey(placeVisitId);
+    Optional<PlaceVisitModel> optionalPlace = Optional.empty();
+    try {
+      Entity placeEntity = datastore.get(placeKey);
+      String mark = (String) placeEntity.getProperty(USER_MARK_PROPERTY_NAME);
+      PlaceVisitModel.UserMark userMark = PlaceVisitModel.UserMark.valueOf(mark);
+      optionalPlace = Optional.of(
+          PlaceVisitModel.builder()
+              .setId((String) placeEntity.getProperty(ID_PROPERTY_NAME))
+              .setPlaceName((String) placeEntity.getProperty(PLACE_NAME_PROPERTY_NAME))
+              .setPlacesApiPlaceId((String) placeEntity.getProperty(PLACE_API_ID_PROPERTY_NAME))
+              .setTripId((String) placeEntity.getProperty(TRIP_ID_PROPERTY_NAME))
+              .setUserMark(userMark)
+              .build());
+
+    } catch (EntityNotFoundException e) {
+    }
+    return optionalPlace;
+  }
+
+  public PlaceVisitModel updateUserMarkOrAddPlaceVisit(
+      PlaceVisitModel placeVisit, PlaceVisitModel.UserMark newStatus) {
+    Key placeKey = KeyFactory.stringToKey(placeVisit.id());
+    try {
+      Entity placeEntity = datastore.get(placeKey);
+      placeEntity.setProperty(USER_MARK_PROPERTY_NAME, newStatus.toString());
+      datastore.put(placeEntity);
+      placeVisit.toBuilder().setUserMark(newStatus);
+
+    } catch (EntityNotFoundException e) {
```

Comment:
This doesn't satisfy the contract of the interface: we're supposed to insert the given place if it doesn't exist here.

Now might be a good time to reconsider the interface semantics and any changes we'd like to make to it. You may want to discuss it with @cli1903. This is why I mentioned tweaking the interface method to _only_ accept a `PlaceVisitModel` and updating it in-place (`updatePlaceVisit` or perhaps `updateOrInsertPlaceVisit`). This makes the API semantics much clearer, as there's only one possible implementation.